nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/ScanKudu.java

```diff
@@ -0,0 +1,377 @@
+
+        FlowFile openedFF = handler.getFlowFile();
+        if (openedFF != null) {
+            finalizeFlowFile(session, openedFF, tableName, handler.getRecordsCount(), null);
+        }
+
+        session.transfer(flowFile, REL_ORIGINAL);
+        session.commit();
+
+    }
+
+    /**
+     * Result Handler for Scan operation
+     */
+    private class ScanKuduResultHandler implements ResultHandler {
+
+        final private ProcessSession session;
+        final private FlowFile origFF;
+        final private AtomicReference<Long> rowsPulledHolder;
+        final private AtomicReference<Long> ffCountHolder;
+        final private String tableName;
+        final private Integer bulkSize;
+        private FlowFile flowFile = null;
+        final private byte[] JSON_ARRAY_DELIM = ",\n".getBytes();
+
+        private boolean handledAny = false;
+
+        ScanKuduResultHandler(final ProcessSession session,
+                              final FlowFile origFF, final AtomicReference<Long> rowsPulledHolder, final AtomicReference<Long> ffCountHolder,
+                              final String tableName, final Integer bulkSize){
+            this.session = session;
+            this.rowsPulledHolder = rowsPulledHolder;
+            this.ffCountHolder = ffCountHolder;
+            this.tableName = tableName;
+            this.bulkSize = bulkSize == null ? 0 : bulkSize;
+            this.origFF = origFF;
+
+        }
+
+        @Override
+        public void handle(final RowResult resultCells) {
+
+            long rowsPulled = rowsPulledHolder.get();
+            long ffUncommittedCount = ffCountHolder.get();
+
+            try{
+                if (flowFile == null){
+                    flowFile = initNewFlowFile(session, origFF, tableName);
+                    ffUncommittedCount++;
+                }
+
+                flowFile = session.append(flowFile, (out) -> {
+                    if (rowsPulledHolder.get() > 0){
+                        out.write(JSON_ARRAY_DELIM);
+                    }
+                    final String json = convertToJson(resultCells);
+                    out.write(json.getBytes());
+                });
+                handledAny = true;
+
+            }catch(Exception e){
+                throw new RuntimeException(e);
+            }
+
+            rowsPulled++;
+
+            // bulkSize controls number of records per flow file.
+            if (bulkSize > 0 && rowsPulled >= bulkSize) {
+
+                finalizeFlowFile(session, flowFile, tableName, rowsPulled, null);
+                flowFile = null;
+                rowsPulledHolder.set(0L);
+
+                // we could potentially have a huge number of rows. If we get to batchSize, go ahead and commit the
+                // session so that we can avoid buffering tons of FlowFiles without ever sending any out.
+                if (getBatchSize()>0 && ffUncommittedCount*bulkSize > getBatchSize()) {
```

Comment:
This is a bit confusing. How are `batchSize` and `bulkSize` related?
Isn't `bulkSize` set to `batchSize` in the constructor?